packages/runner/src/fixture.ts

```diff
@@ -78,49 +78,45 @@ export function withFixtures(fn: Function, testContext?: TestContext) {
     if (!pendingFixtures.length)
       return fn(context)
 
-    let cursor = 0
-
-    return new Promise((resolve, reject) => {
-      async function use(fixtureValue: any) {
-        const fixture = pendingFixtures[cursor++]
-        context![fixture.prop] = fixtureValue
-
-        if (!fixtureValueMap.has(fixture)) {
-          fixtureValueMap.set(fixture, fixtureValue)
-          cleanupFnArray.unshift(() => {
-            fixtureValueMap.delete(fixture)
+    async function resolveFixtures() {
+      for (const fixture of pendingFixtures) {
+        // fixture could be already initialized during "before" hook
+        if (fixtureValueMap.has(fixture))
+          continue
+
+        let fixtureValue: unknown
+        if (fixture.isFn) {
+          // wait for `use` call to extract fixture value
+          let isFixtureTeardown = false
+          fixtureValue = await new Promise((resolveUseArg, rejectUseArg) => {
+            fixture.value(context, (useArg: unknown) => {
+              resolveUseArg(useArg)
+              isFixtureTeardown = true
+              // continue fixture function during cleanup
+              return new Promise<void>((resolveUseReturn) => {
+                cleanupFnArray.push(resolveUseReturn)
+              })
+            }).catch((e: unknown) => {
+              // re-throw if error is thrown during fixture teardown
+              if (isFixtureTeardown)
+                throw e
+              // otherwise it as a failure of a test calling this fixture
+              rejectUseArg(e)
+            })
           })
         }
-
-        if (cursor < pendingFixtures.length) {
-          await next()
-        }
         else {
-          // When all fixtures setup, call the test function
-          try {
-            resolve(await fn(context))
-          }
-          catch (err) {
-            reject(err)
-          }
-          return new Promise<void>((resolve) => {
-            cleanupFnArray.push(resolve)
-          })
+          fixtureValue = fixture.value
         }
+        context![fixture.prop] = fixtureValue
+        fixtureValueMap.set(fixture, fixtureValue)
+        cleanupFnArray.unshift(() => {
+          fixtureValueMap.delete(fixture)
+        })
       }
+    }
 
-      async function next() {
-        const fixture = pendingFixtures[cursor]
-        const { isFn, value } = fixture
-        if (fixtureValueMap.has(fixture))
-          return use(fixtureValueMap.get(fixture))
-        else
-          return isFn ? value(context, use) : use(value)
-      }
-
-      const setupFixturePromise = next().catch(reject)
-      cleanupFnArray.unshift(() => setupFixturePromise)
```

Comment:
Maybe the issue could be simply this and removing this might fix https://github.com/vitest-dev/vitest/issues/4669